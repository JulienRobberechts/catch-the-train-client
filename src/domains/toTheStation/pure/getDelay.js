import moment from "moment";
import getDelayStatus from "./getDelayStatus";

export default function getDelay({
  nowTime,
  targetTime,
  travelDuration,
  waitingDuration
}) {
  const targetDuration = moment.duration(targetTime.diff(nowTime));

  // positive will be early and negative late
  const delayDuration = targetDuration.clone();
  delayDuration.subtract(travelDuration);
  delayDuration.subtract(waitingDuration);

  return {
    targetDuration,
    delayDuration
  };
}
