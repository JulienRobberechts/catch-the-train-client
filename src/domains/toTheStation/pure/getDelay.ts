import moment from "moment";

interface GetDelayParams {
  nowTime: moment.Moment;
  departureTime: moment.Moment;
  travelDuration: moment.Duration;
  accessDuration: moment.Duration;
}

export default function getDelay({
  nowTime,
  departureTime,
  travelDuration,
  accessDuration,
}: GetDelayParams) {
  const departureDuration = moment.duration(departureTime.diff(nowTime));

  // positive will be early and negative late
  const delayDuration = departureDuration.clone();
  delayDuration.subtract(travelDuration);
  delayDuration.subtract(accessDuration);

  return {
    departureDuration,
    delayDuration,
  };
}
