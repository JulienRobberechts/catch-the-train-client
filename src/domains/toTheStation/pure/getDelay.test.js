import moment from "moment";
import getDelay from "./getDelay";
import each from "jest-each";

describe("train helpers", () => {
  describe("getDelay", () => {
    each([
      [
        "early 1min",
        {
          now: "2020-03-10T09:00:00+01:00",
          train: "2020-03-10T09:11:00+01:00",
          travel: "00:08:00",
          waiting: "00:02:00",
          expectedTargetDuration: 60 * 11,
          expectedDelayDuration: 60 * 1
        }
      ],
      [
        "on time 0s",
        {
          now: "2020-03-10T09:01:00+01:00",
          train: "2020-03-10T09:11:00+01:00",
          travel: "00:08:00",
          waiting: "00:02:00",
          expectedTargetDuration: 60 * 10,
          expectedDelayDuration: 60 * 0
        }
      ],
      [
        "on time -15s",
        {
          now: "2020-03-10T09:01:15+01:00",
          train: "2020-03-10T09:11:00+01:00",
          travel: "00:08:00",
          waiting: "00:02:00",
          expectedTargetDuration: 60 * 10 - 15,
          expectedDelayDuration: 60 * 0 - 15
        }
      ],
      [
        "late -1min",
        {
          now: "2020-03-10T09:02:00+01:00",
          train: "2020-03-10T09:11:00+01:00",
          travel: "00:08:00",
          waiting: "00:02:00",
          expectedTargetDuration: 60 * 9,
          expectedDelayDuration: 60 * -1
        }
      ]
    ]).test(
      "should calculate delay for %s",
      (
        _name,
        {
          now,
          train,
          travel,
          waiting,
          expectedTargetDuration,
          expectedDelayDuration
        }
      ) => {
        const nowTime = moment.parseZone(now);
        const targetTime = moment.parseZone(train);
        const travelDuration = moment.duration(travel);
        const waitingDuration = moment.duration(waiting);

        const result = getDelay({
          nowTime,
          targetTime,
          travelDuration,
          waitingDuration
        });

        expect(result).toEqual(expect.anything());
        const { targetDuration, delayDuration } = result;
        expect(targetDuration.asSeconds()).toEqual(expectedTargetDuration);
        expect(delayDuration.asSeconds()).toEqual(expectedDelayDuration);
      }
    );
  });
});
