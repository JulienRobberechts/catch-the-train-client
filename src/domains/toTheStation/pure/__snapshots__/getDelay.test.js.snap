// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`train helpers getDelay should identify an late delay 1`] = `
Object {
  "delayDuration": "PT1M",
  "targetDuration": "PT11M",
}
`;
